cmd/microcloud/status.go

```diff
@@ -0,0 +1,451 @@
+	// Systems that exist in other clusters but not in MicroCloud.
+	unmanagedSystems := map[types.ServiceType]map[string]bool{}
+
+	// Systems that exist in MicroCloud, but not other clusters.
+	orphanedSystems := map[types.ServiceType]map[string]bool{}
+
+	// Services that are uninitialized on a system.
+	uninstalledServices := map[types.ServiceType][]string{}
+
+	// Services undergoing schema/API upgrades.
+	upgradingServices := map[types.ServiceType]bool{}
+
+	// Systems that are offline on at least one service.
+	offlineSystems := map[string]bool{}
+
+	osdsConfigured := false
+	clusterSize := 0
+	osdCount := 0
+
+	for _, s := range statuses {
+		if s.Name == name {
+			clusterSize = len(s.Clusters[types.MicroCloud])
+			for service, clusterMembers := range s.Clusters {
+				for _, member := range clusterMembers {
+					if member.Status == microTypes.MemberNeedsUpgrade || member.Status == microTypes.MemberUpgrading {
+						upgradingServices[service] = true
+					} else if member.Status != microTypes.MemberOnline {
+						offlineSystems[member.Name] = true
+					}
+				}
+			}
+		}
+
+		osdCount = osdCount + len(s.OSDs)
+		allServices := []types.ServiceType{types.LXD, types.MicroCeph, types.MicroOVN, types.MicroCloud}
+		cloudMembers := make(map[string]bool, len(s.Clusters[types.MicroCloud]))
+		for _, member := range s.Clusters[types.MicroCloud] {
+			cloudMembers[member.Name] = true
+		}
+
+		for _, service := range allServices {
+			if len(s.Clusters[service]) == 0 {
+				if uninstalledServices[service] == nil {
+					uninstalledServices[service] = []string{}
+				}
+
+				uninstalledServices[service] = append(uninstalledServices[service], s.Name)
+			}
+
+			if service == types.MicroCloud || s.Name != name {
+				continue
+			}
+
+			for _, member := range s.Clusters[service] {
+				if !cloudMembers[member.Name] {
+					if unmanagedSystems[service] == nil {
+						unmanagedSystems[service] = map[string]bool{}
+					}
+
+					unmanagedSystems[service][member.Name] = true
+				}
+			}
+
+			if len(s.Clusters[service]) > 0 {
+				clusterMap := make(map[string]bool, len(s.Clusters[service]))
+				for _, member := range s.Clusters[service] {
+					clusterMap[member.Name] = true
+				}
+
+				for name := range cloudMembers {
+					if !clusterMap[name] {
+						if orphanedSystems[service] == nil {
+							orphanedSystems[service] = map[string]bool{}
+						}
+
+						orphanedSystems[service][name] = true
+					}
+				}
+			}
+		}
+
+		if osdCount > 0 && len(s.Clusters[types.MicroCeph]) > 0 {
+			osdsConfigured = true
+		}
+	}
+
+	// Format the actual warnings based on the collected data.
+	warnings := Warnings{}
+	if clusterSize < 3 {
+		tmpl := tui.Fmt{Arg: "%s: %d systems are required for effective fault tolerance"}
+		msg := tui.Printf(tmpl,
+			tui.Fmt{Color: tui.Red, Arg: "Reliability risk", Bold: true},
+			tui.Fmt{Color: tui.Bright, Arg: 3, Bold: true},
+		)
+
+		warnings = append(warnings, Warning{Level: Error, Message: msg})
```

Comment:
Should this be `Level: Warning` instead?

Reply:
My assumption so far was that we consider falling short of the HA targets to be basically "invalid". We should discuss this further.

Reply:
See https://github.com/canonical/microcloud/pull/384#discussion_r1795714711. I am happy if this stays as error then.